src/storage/segcache/src/builder.rs

```diff
@@ -0,0 +1,172 @@
+            overflow_factor: 0.0,
+            segments_builder: SegmentsBuilder::default(),
+        }
+    }
+}
+
+impl Builder {
+    /// Specify the hash power, which limits the size of the hashtable to 2^N
+    /// entries. 1/8th of these are used for metadata storage, meaning that the
+    /// total number of items which can be held in the cache is limited to
+    /// `7 * 2^(N - 3)` items. The hash table will have a total size of
+    /// `2^(N + 3)` bytes.
+    ///
+    /// ```
+    /// use segcache::SegCache;
+    ///
+    /// // create a cache with a small hashtable that has room for ~114k items
+    /// // without using any overflow buckets.
+    /// let cache = SegCache::builder().power(17).build();
+    ///
+    /// // create a cache with a larger hashtable with room for ~1.8M items
+    /// let cache = SegCache::builder().power(21).build();
+    /// ```
+    pub fn power(mut self, power: u8) -> Self {
+        assert!(power >= 3, "power must be at least 3");
+        self.power = power;
+        self
+    }
+
+    /// Specify an overflow factor which is used to scale the hashtable and
+    /// provide additional capacity for chaining item buckets. A factor of 1.0
+    /// will result in a hash table that is 100% larger.
+    ///
+    /// ```
+    /// use segcache::SegCache;
+    ///
+    /// // create a cache with a hashtable with room for ~228k items, which is
+    /// // about the same as using a power of 18, but is more tolerant of hash
+    /// // collisions
+    /// let cache = SegCache::builder()
+    ///     .power(17)
+    ///     .overflow_factor(1.0)
+    ///     .build();
+    ///
+    /// // smaller overflow factors may be specified, meaning only some buckets
+    /// // can ever be chained
+    /// let cache = SegCache::builder()
+    ///     .power(17)
+    ///     .overflow_factor(0.2)
+    ///     .build();
+    /// ```
+    pub fn overflow_factor(mut self, percent: f64) -> Self {
+        self.overflow_factor = percent;
+        self
+    }
+
+    /// Specify the total number of bytes to be used for heap storage of items.
+    /// This includes, key, value, and per-item overheads.
+    ///
+    /// ```
+    /// use segcache::SegCache;
+    ///
+    /// const MB: usize = 1024 * 1024;
+    ///
+    /// // create a cache with a 64MB heap
+    /// let cache = SegCache::builder().heap_size(64 * MB).build();
+    ///
+    /// // create a cache with a 256MB heap
+    /// let cache = SegCache::builder().heap_size(256 * MB).build();
+    /// ```
+    pub fn heap_size(mut self, bytes: usize) -> Self {
+        self.segments_builder = self.segments_builder.heap_size(bytes);
+        self
+    }
+
+    /// Specify the segment size for item storage. The largest item which can be
+    /// held is `size - 5` bytes for builds without the `debug` or `magic` build
+    /// features enabled. Smaller segment sizes reduce the number of items which
+    /// would be evicted/expired at one time, at the cost of additional memory
```

Comment:
In the comment/doc, I think we should still point out that segment size is an upper bound for single item size. This is obvious for anybody familiar with (memcached) slabs, but may not be obvious generally. Another "pro-tip" is we often don't want to config the segment size to be too close to the largest item size *if* item sizes tend to vary within a broad spectrum, because this will leave more bytes stranded due to internal fragmentation. Instead, we can advise users to config segment size to be at least a small multiple of the largest item size in that case.